File: packages/skillet-eval-python/src/skill_eval/runner/context.py
from __future__ import annotations

import json

from skill_eval.providers.types import Message
from skill_eval.utils.string import truncate

CHARS_PER_TOKEN = 4
MAX_CONTEXT_TOKENS = 80_000
MAX_CONTEXT_CHARS = MAX_CONTEXT_TOKENS * CHARS_PER_TOKEN
COMPACT_MIN_LENGTH = 200
PROTECTED_TAIL_MESSAGES = 6
SNIPPET_MAX_LENGTH = 80
TOOL_ARG_DISPLAY_MAX = 60


def message_size(m: Message) -> int:
    size = len(m.content) if m.content else 0
    if m.tool_calls:
        for tc in m.tool_calls:
            size += len(json.dumps(tc.arguments))
    return size


def total_payload_size(system: str, messages: list[Message]) -> int:
    return len(system) + sum(message_size(m) for m in messages)


def compact_messages(messages: list[Message], system: str) -> None:
    if total_payload_size(system, messages) <= MAX_CONTEXT_CHARS:
        return

    placeholder = "[truncated]"

    for i in range(1, len(messages) - PROTECTED_TAIL_MESSAGES):
        if total_payload_size(system, messages) <= MAX_CONTEXT_CHARS:
            break
        msg = messages[i]
        if msg.content and len(msg.content) > COMPACT_MIN_LENGTH:
            msg.content = placeholder
        if msg.tool_calls:
            for tc in msg.tool_calls:
                for key, val in list(tc.arguments.items()):
                    if isinstance(val, str) and len(val) > COMPACT_MIN_LENGTH:
                        tc.arguments[key] = placeholder


def extract_snippet(content: str) -> str | None:
    if not content or not isinstance(content, str):
        return None
    for line in content.split("\n"):
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#") or trimmed.startswith("```") or trimmed.startswith("<"):
            continue
        return truncate(trimmed, SNIPPET_MAX_LENGTH)
    return None


def summarize_tool_args(name: str, args: dict[str, object]) -> str:
    if name == "bash" and isinstance(args.get("command"), str):
        return truncate(args["command"], TOOL_ARG_DISPLAY_MAX)
    if name in ("read_file", "write_file", "list_directory") and isinstance(args.get("path"), str):
        return args["path"]
    return truncate(json.dumps(args), TOOL_ARG_DISPLAY_MAX)


def debug_payload_log(
    step: int,
    messages: list[Message],
    system: str,
) -> tuple[int, list[str]]:
    pre_size = total_payload_size(system, messages)
    breakdown = []
    for i, m in enumerate(messages):
        s = message_size(m)
        if s > 1000:
            breakdown.append(f"  [{i}] {m.role} {s / 1024:.1f}KB")
    return pre_size, breakdown

File: packages/skillet/src/eval/runner/context.ts
import type { Message } from "../providers/types.js";
import { truncate } from "../utils/string.js";

const CHARS_PER_TOKEN = 4;
const MAX_CONTEXT_TOKENS = 80_000;
const MAX_CONTEXT_CHARS = MAX_CONTEXT_TOKENS * CHARS_PER_TOKEN;
const COMPACT_MIN_LENGTH = 200;
const PROTECTED_TAIL_MESSAGES = 6;
const SNIPPET_MAX_LENGTH = 80;
const TOOL_ARG_DISPLAY_MAX = 60;

export function messageSize(m: Message): number {
	let size = m.content?.length ?? 0;
	if (m.toolCalls) {
		for (const tc of m.toolCalls) {
			size += JSON.stringify(tc.arguments).length;
		}
	}
	return size;
}

export function totalPayloadSize(system: string, messages: Message[]): number {
	return system.length + messages.reduce((sum, m) => sum + messageSize(m), 0);
}

/**
 * If total message payload exceeds the context budget, aggressively
 * truncate older messages (oldest first). Protects the last few messages.
 */
export function compactMessages(messages: Message[], system: string) {
	if (totalPayloadSize(system, messages) <= MAX_CONTEXT_CHARS) return;

	const PLACEHOLDER = "[truncated]";

	for (let i = 1; i < messages.length - PROTECTED_TAIL_MESSAGES; i++) {
		if (totalPayloadSize(system, messages) <= MAX_CONTEXT_CHARS) break;
		const msg = messages[i];
		if (msg.content && msg.content.length > COMPACT_MIN_LENGTH) {
			msg.content = PLACEHOLDER;
		}
		if (msg.toolCalls) {
			for (const tc of msg.toolCalls) {
				for (const [key, val] of Object.entries(tc.arguments)) {
					if (typeof val === "string" && val.length > COMPACT_MIN_LENGTH) {
						tc.arguments[key] = PLACEHOLDER;
					}
				}
			}
		}
	}
}

/** Extract the first meaningful line from model output as a display snippet. */
export function extractSnippet(content: string): string | undefined {
	if (!content || typeof content !== "string") return undefined;
	for (const line of content.split("\n")) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#") || trimmed.startsWith("```") || trimmed.startsWith("<"))
			continue;
		return truncate(trimmed, SNIPPET_MAX_LENGTH);
	}
	return undefined;
}

/** Produce a compact display string for tool call arguments. */
export function summarizeToolArgs(name: string, args: Record<string, unknown>): string {
	if (name === "bash" && typeof args.command === "string") {
		return truncate(args.command, TOOL_ARG_DISPLAY_MAX);
	}
	if (
		(name === "read_file" || name === "write_file" || name === "list_directory") &&
		typeof args.path === "string"
	) {
		return args.path;
	}
	return truncate(JSON.stringify(args), TOOL_ARG_DISPLAY_MAX);
}

export function debugPayloadLog(
	step: number,
	messages: Message[],
	system: string,
): { preSize: number; breakdown: string[] } {
	const preSize = totalPayloadSize(system, messages);
	const breakdown = messages
		.map((m, i) => {
			const s = messageSize(m);
			return s > 1000 ? `  [${i}] ${m.role} ${(s / 1024).toFixed(1)}KB` : null;
		})
		.filter((line): line is string => line !== null);
	return { preSize, breakdown };
}
